File: src/harness/local_models.cpp
#include "harness/local_models.h"

#include <algorithm>
#include <cstdlib>
#include <filesystem>
#include <string_view>
#include <system_error>

#include "rac/core/rac_platform_adapter.h"

#include "catalog/catalog.h"

namespace wally::harness {
namespace {

namespace fs = std::filesystem;

bool IsWeightFile(const fs::path& path) {
    const std::string ext = path.extension().string();
    return ext == ".gguf" || ext == ".safetensors" || ext == ".onnx" || ext == ".bin";
}

/// Written by the download orchestrator as files land rather than once they all
/// have, so a partial download carries one too. Presence means "this came from
/// a download", not "the download finished".
constexpr std::string_view kManifest = ".rac-manifest.binpb";

}  // namespace

std::vector<LocalModel> LocalModels(const std::string& home) {
    std::vector<LocalModel> models;
    if (home.empty()) {
        return models;
    }

    // Both layouts are real: the SDK's path docs describe
    // {base}/RunAnywhere/Models, and the desktop default base directory already
    // ends in "runanywhere", so models land directly under {base}/Models.
    std::error_code ec;
    fs::path root = fs::path(home) / "RunAnywhere" / "Models";
    if (!fs::is_directory(root, ec)) {
        root = fs::path(home) / "Models";
    }
    if (!fs::is_directory(root, ec)) {
        return models;
    }

    // Every step of the walk takes an error_code, including the increments.
    // Constructing the iterator with one and then advancing it with the
    // range-for's throwing operator++ only looks safe: a directory removed or
    // locked while this runs escapes as a filesystem_error out of what is a
    // best-effort listing, and this is called from the launch path.
    //
    // `walk` guards iteration and stops that level when advancing fails.
    // `probe` is for the status queries, whose false-on-error answer is already
    // the one we want, so it is reset rather than checked.
    constexpr auto kSkip = fs::directory_options::skip_permission_denied;
    std::error_code walk;
    std::error_code probe;

    for (fs::directory_iterator framework(root, kSkip, walk), no_more_frameworks;
         !walk && framework != no_more_frameworks; framework.increment(walk)) {
        if (!framework->is_directory(probe)) {
            probe.clear();
            continue;
        }
        std::error_code entries;
        for (fs::directory_iterator entry(framework->path(), kSkip, entries), no_more_entries;
             !entries && entry != no_more_entries; entry.increment(entries)) {
            if (!entry->is_directory(probe)) {
                probe.clear();
                continue;
            }
            std::string weights;
            bool manifest = false;
            std::int64_t bytes = 0;
            std::error_code files;
            for (fs::recursive_directory_iterator file(entry->path(), kSkip, files),
                 no_more_files;
                 !files && file != no_more_files; file.increment(files)) {
                if (!file->is_regular_file(probe)) {
                    probe.clear();
                    continue;
                }
                // A file that vanished between the listing and the stat reports
                // an error and an unspecified size; adding that unchecked cast
                // a -1 into a wildly wrong total.
                const std::uintmax_t size = file->file_size(probe);
                if (!probe) {
                    bytes += static_cast<std::int64_t>(size);
                }
                probe.clear();
                if (file->path().filename() == kManifest) {
                    manifest = true;
                } else if (weights.empty() && IsWeightFile(file->path())) {
                    weights = file->path().string();
                }
            }
            // A directory holding weights but no manifest was placed by hand.
            // It is still loadable, so it still counts.
            if (!manifest && weights.empty()) {
                continue;
            }
            models.push_back({entry->path().filename().string(),
                              framework->path().filename().string(), entry->path().string(),
                              weights, bytes});
        }
    }

    std::sort(models.begin(), models.end(),
              [](const LocalModel& a, const LocalModel& b) { return a.id < b.id; });
    return models;
}

std::int64_t LocalContextSize(const std::string& model_id) {
    constexpr std::int64_t kFloor = 8192;
    constexpr std::int64_t kGiB = 1024LL * 1024 * 1024;

    // Physical memory decides the tier. On Apple Silicon this is the unified
    // pool the GPU draws from too, which is why it stands in for VRAM here.
    std::int64_t tier = kFloor;
    const rac_platform_adapter_t* adapter = rac_get_platform_adapter();
    rac_memory_info_t memory{};
    if (adapter != nullptr && adapter->get_memory_info != nullptr &&
        adapter->get_memory_info(&memory, adapter->user_data) == RAC_SUCCESS &&
        memory.total_bytes > 0) {
        const std::int64_t total = static_cast<std::int64_t>(memory.total_bytes);
        if (total >= 48 * kGiB) {
            tier = 65536;
        } else if (total >= 24 * kGiB) {
            tier = 32768;
        } else if (total >= 12 * kGiB) {
            tier = 16384;
        }
    }

    // The model's own window caps it: asking llama.cpp for more than the model
    // was trained on stretches RoPE and degrades every answer. 0 means the
    // catalog does not know, and a model that is not in the catalog at all
    // (an hf.co ref, a hand-placed folder) gets the tier as is.
    std::int64_t window = tier;
    if (const catalog::CatalogEntry* entry = catalog::find(model_id)) {
        if (entry->context_length > 0) {
            window = std::min(tier, static_cast<std::int64_t>(entry->context_length));
        }
    }
    return window;
}

std::int64_t LocalOutputSize(std::int64_t context_window) {
    return context_window > 1 ? std::min<std::int64_t>(4096, context_window / 4) : 0;
}

}  // namespace wally::harness

File: src/harness/local_models.rs
//! Models downloaded on this machine, for coding tools (port of
//! src/harness/local_models.cpp).

use std::fs;
use std::path::{Path, PathBuf};

use crate::catalog;
use crate::sys;

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct LocalModel {
    pub id: String,
    pub framework: String,
    pub dir: String,
    pub path: String,
    pub bytes: i64,
}

fn is_weight_file(path: &Path) -> bool {
    matches!(
        path.extension().and_then(|e| e.to_str()),
        Some("gguf") | Some("safetensors") | Some("onnx") | Some("bin")
    )
}

/// Written by the download orchestrator as files land rather than once they all
/// have, so a partial download carries one too. Presence means "this came from
/// a download", not "the download finished".
const MANIFEST: &str = ".rac-manifest.binpb";

/// Sums bytes and finds the manifest/first weight file for one model directory,
/// walking recursively. Best-effort: a file that vanishes mid-walk is skipped
/// rather than failing the whole listing.
fn scan_model_dir(dir: &Path) -> (String, bool, i64) {
    let mut weights = String::new();
    let mut manifest = false;
    let mut bytes: i64 = 0;
    let mut stack = vec![dir.to_path_buf()];
    while let Some(current) = stack.pop() {
        let entries = match fs::read_dir(&current) {
            Ok(e) => e,
            Err(_) => continue,
        };
        for entry in entries.flatten() {
            let path = entry.path();
            let file_type = match entry.file_type() {
                Ok(t) => t,
                Err(_) => continue,
            };
            // `file_type()` describes the link itself, not what it points at,
            // so a symlinked weight file would otherwise look like neither a
            // directory nor a regular file and vanish from the walk.
            // `fs::metadata` follows the link to learn the target's real
            // type and size -- `DirEntry::metadata()` (used below for a
            // plain file) does not follow a symlink either, and lstat's
            // reported size for one is the length of the target path
            // string, not the target's content, so the resolved metadata
            // fetched here is reused for the byte count too.
            let target_metadata = if file_type.is_symlink() {
                match fs::metadata(&path) {
                    Ok(target) => Some(target),
                    // Broken link, or a stat that failed: skip, same as this
                    // function's other best-effort cases.
                    Err(_) => continue,
                }
            } else {
                None
            };
            let is_dir = match &target_metadata {
                Some(target) if target.is_dir() => true,
                Some(target) if target.is_file() => false,
                // A symlink to a special file (socket, device, ...): skip.
                Some(_) => continue,
                None if file_type.is_dir() => true,
                None if file_type.is_file() => false,
                None => continue,
            };
            if is_dir {
                stack.push(path);
                continue;
            }
            // A file that vanished between the listing and the stat is skipped
            // rather than counted as size 0, which would otherwise silently
            // undercount the total.
            let size = match &target_metadata {
                Some(target) => Some(target.len()),
                None => entry.metadata().ok().map(|m| m.len()),
            };
            if let Some(size) = size {
                bytes += size as i64;
            }
            if path.file_name().and_then(|n| n.to_str()) == Some(MANIFEST) {
                manifest = true;
            } else if weights.is_empty() && is_weight_file(&path) {
                weights = path.to_string_lossy().into_owned();
            }
        }
    }
    (weights, manifest, bytes)
}

/// Models present under `home` right now, found by walking the storage tree.
///
/// Walking rather than asking the registry: this only has to answer "is there
/// something here the local server can open", and the walk says that about a
/// model placed by hand as readily as one that was downloaded.
pub fn local_models(home: &str) -> Vec<LocalModel> {
    let mut models = Vec::new();
    if home.is_empty() {
        return models;
    }

    // Both layouts are real: the SDK's path docs describe
    // {base}/RunAnywhere/Models, and the desktop default base directory already
    // ends in "runanywhere", so models land directly under {base}/Models.
    let mut root: PathBuf = Path::new(home).join("RunAnywhere").join("Models");
    if !root.is_dir() {
        root = Path::new(home).join("Models");
    }
    if !root.is_dir() {
        return models;
    }

    let Ok(frameworks) = fs::read_dir(&root) else {
        return models;
    };
    for framework_entry in frameworks.flatten() {
        let Ok(framework_type) = framework_entry.file_type() else {
            continue;
        };
        if !framework_type.is_dir() {
            continue;
        }
        let framework_name = framework_entry.file_name().to_string_lossy().into_owned();
        let Ok(entries) = fs::read_dir(framework_entry.path()) else {
            continue;
        };
        for entry in entries.flatten() {
            let Ok(entry_type) = entry.file_type() else {
                continue;
            };
            if !entry_type.is_dir() {
                continue;
            }
            let entry_path = entry.path();
            let (weights, manifest, bytes) = scan_model_dir(&entry_path);
            // A directory holding weights but no manifest was placed by hand.
            // It is still loadable, so it still counts.
            if !manifest && weights.is_empty() {
                continue;
            }
            models.push(LocalModel {
                id: entry.file_name().to_string_lossy().into_owned(),
                framework: framework_name.clone(),
                dir: entry_path.to_string_lossy().into_owned(),
                path: weights,
                bytes,
            });
        }
    }

    models.sort_by(|a, b| a.id.cmp(&b.id));
    models
}

/// The context window a local server is started with for `model_id`, in
/// tokens. Sized from this machine's memory in tiers (8k / 16k / 32k / 64k),
/// because a coding agent's first request is a 15k-token system prompt and a
/// fixed 8k window rejected it outright. Capped at the model's own window when
/// the catalog knows it (even when that is below the 8k memory tier). One
/// function, so the server, the picker's declared limits, and the shim all
/// quote the same number.
pub fn local_context_size(model_id: &str) -> i64 {
    const FLOOR: i64 = 8192;
    const GIB: i64 = 1024 * 1024 * 1024;

    // Physical memory decides the tier. On Apple Silicon this is the unified
    // pool the GPU draws from too, which is why it stands in for VRAM here.
    let mut tier = FLOOR;
    // SAFETY: rac_get_platform_adapter() returns either null or a pointer the
    // SDK keeps valid for the process lifetime; the function pointer slot is
    // checked for null before it is called.
    unsafe {
        let adapter = sys::rac_get_platform_adapter();
        if !adapter.is_null() {
            if let Some(get_memory_info) = (*adapter).get_memory_info {
                let mut memory = sys::rac_memory_info_t {
                    total_bytes: 0,
                    available_bytes: 0,
                    used_bytes: 0,
                };
                if get_memory_info(&mut memory, (*adapter).user_data) == sys::SUCCESS
                    && memory.total_bytes > 0
                {
                    let total = memory.total_bytes as i64;
                    if total >= 48 * GIB {
                        tier = 65536;
                    } else if total >= 24 * GIB {
                        tier = 32768;
                    } else if total >= 12 * GIB {
                        tier = 16384;
                    }
                }
            }
        }
    }

    // The model's own window caps it: asking llama.cpp for more than the model
    // was trained on stretches RoPE and degrades every answer. 0 means the
    // catalog does not know, and a model that is not in the catalog at all
    // (an hf.co ref, a hand-placed folder) gets the tier as is.
    let mut window = tier;
    if let Some(entry) = catalog::find(model_id) {
        if entry.context_length > 0 {
            window = tier.min(entry.context_length as i64);
        }
    }
    window
}

/// Leave most of a local context for the coding prompt and conversation.
/// Zero means no local context was configured (a hosted endpoint).
pub fn local_output_size(context_window: i64) -> i64 {
    if context_window > 1 {
        4096.min(context_window / 4)
    } else {
        0
    }
}
